clang/lib/Frontend/CompilerInvocation.cpp

```diff
@@ -4675,7 +4676,7 @@ bool CompilerInvocation::CreateFromArgsImpl(
   // FIXME: We shouldn't have to pass the DashX option around here
   InputKind DashX = Res.getFrontendOpts().DashX;
   ParseTargetArgs(Res.getTargetOpts(), Args, Diags);
-  llvm::Triple T(Res.getTargetOpts().Triple);
+  llvm::Triple T(llvm::Triple::normalize(Res.getTargetOpts().Triple));
```

Comment:
> Looks like part of this change is to normalize `Res.getTargetOpts().Triple` in many places where it is used.
> 
> Would it be possible to arrange it so that this is normalized in one place earlier on, so all users of `getTargetOpts()` can assume that it is normalized already?

I believe the triple accessible using the expression `getTargetOpts().Triple` denotes the target options specified for the compilation. It seems appropriate to leave it unmodified (un-normalized) for reference or other purposes during compilation.

Reply:
IIUC `getTargetOpts().Triple` should *already* be normalized. The definition of the option includes a "Normalizer", like so:
```
def triple : Separate<["-"], "triple">,
  HelpText<"Specify target triple (e.g. i686-apple-darwin9)">,
  MarshallingInfoString<TargetOpts<"Triple">, "llvm::Triple::normalize(llvm::sys::getDefaultTargetTriple())">,
  AlwaysEmit, Normalizer<"normalizeTriple">;
```

So I think we shouldn't need the call to normalizer in CompilerInvocation at all, and we should look at the other places we're adding it to make sure they all make sense.